### PR TITLE
feat(behavior_path_side_shift): add drivable area check to prevent lane departure

### PR DESCRIPTION
**Pull request by k-uta**
## Description

This PR adds a lane boundary check to the `side_shift` module to prevent the ego vehicle from departing the drivable area when a lateral offset is requested.

Previously, the module executed the requested shift amount unconditionally, which could cause unsafe lane departures. With this change, the module calculates the maximum allowable lateral offset based on the distance to the lanelet boundaries and clamps the requested value so that the vehicle remains within the safe area.

## Changes

- **Added `drivable_area_utils.hpp`**: Introduced helper structs (`AdjacentLaneInfo`, `LaneLimitInfo`, `LanePositionResult`, `LaneCheckContext`) and free functions (`determineLanePosition`, `updateLaneLimitsForOutsidePoint`, `updateLaneLimitsForInsidePoint`, `clampOffsetToLimits`, `guardAgainstSnapBack`) to encapsulate the boundary-check logic.
- **Added `calcMaxLateralOffset` method**: Calculates the safe shift limit based on the vehicle width, configured margin, and distance to lane boundaries. The check range extends from the shift start point to the end of the shift plus a buffer.
- **Added `getAdjacentLaneInfo` method**: Retrieves information about same-direction neighbor lanes when Mode 2 is enabled.
- **Added check modes**:
  - **Mode 0 (Disabled)**: No boundary check; behaves the same as before.
  - **Mode 1 (Current lane only)**: Limits the offset to the current lane boundaries.
  - **Mode 2 (Current + neighbor lanes)**: Extends the drivable area into adjacent same-direction lanes.
- **Added parameters**: `drivable_area_check_mode` and `min_drivable_area_margin`.
- **Safety improvements**: Added guards to abort shifting when the reference path is too short or the shift span is insufficient.

### ROS Parameters

The following parameters have been added to `side_shift.param.yaml`:

| Change Type | Parameter Name | Type | Default | Description |
| :--- | :--- | :--- | :--- | :--- |
| Added | `drivable_area_check_mode` | int | `1` | `0`: No check / `1`: Current lane only / `2`: Current + same-direction neighbor lanes |
| Added | `min_drivable_area_margin` | double | `0.3` | Minimum margin [m] between the vehicle footprint edge and the lane boundary |

## Effects on System Behavior

- **Mode 0**: Behaves identically to the previous implementation; the vehicle shifts by the exact requested amount.
- **Mode 1**: The vehicle shifts toward the target but stops before crossing the boundary of the current lane.
- **Mode 2**: The vehicle can cross the lane marking into an adjacent lane (if one exists in the same direction) but stops before leaving the combined drivable area.

## How Was This PR Tested?

1. Launch the planning simulator.
2. Set `drivable_area_check_mode` to `1` or `2`.
3. Publish a large lateral offset (e.g., −5.0 m) that exceeds the lane width and verify the vehicle stays within the boundary.

**Command:**
```bash
sec=$(($(date +%s))); ros2 topic pub /planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/input/lateral_offset tier4_planning_msgs/msg/LateralOffset "{stamp: {sec: $sec, nanosec: 999999999}, lateral_offset: -5.0}" -1

### CODE REVIEW

File: planning/behavior_path_planner/autoware_behavior_path_side_shift_module/config/side_shift.param.yaml [hunk 1]
```diff
@@ -7,4 +7,6 @@
       min_shifting_distance: 5.0
       min_shifting_speed: 5.56
       shift_request_time_limit: 1.0
+      drivable_area_check_mode: 1 # 0: no check, 1: current lane, 2: current lane + neighbor lane
```
[the remark this answers is not in the document]
Comment (k-uta) [reply planning/behavior_path_planner/autoware_behavior_path_side_shift_module/config/side_shift.param.yaml]: Good catch. It was a typo in the description (likely confused with `min_drivable_area_margin`). I have updated the PR description to use the correct parameter name `drivable_area_check_mode`.

File: planning/behavior_path_planner/autoware_behavior_path_side_shift_module/src/scene.cpp [hunk 2]
```diff
@@ -415,48 +514,223 @@ BehaviorModuleOutput SideShiftModule::adjustDrivableArea(const ShiftedPath & pat
...
 
-  PathWithLaneId extended_path{};
-  {
-    extended_path.points.insert(
-      extended_path.points.end(), prev_reference.points.begin() + clip_idx,
-      prev_reference.points.begin() + prev_ego_idx);
+  const double dist_to_left = lanelet::geometry::distance2d(left_bound, target_point);
+  const double dist_to_right = lanelet::geometry::distance2d(right_bound, target_point);
+
+  const double extra_buffer = 0.05;
```
[the remark this answers is not in the document]
Comment (k-uta) [reply planning/behavior_path_planner/autoware_behavior_path_side_shift_module/src/scene.cpp]: Thanks for flagging this! 

I've replaced the local magic number with a named `constexpr kLaneBoundaryDiscretizationBuffer` at the namespace scope, accompanied by a detailed comment explaining its derivation. 

Specifically, the lanelet boundary polyline discretization error is bounded by `segment_length² / (8 × radius)`, which is ~0.013 m for typical maps. The 0.05 m value provides ~4× headroom to safely absorb floating-point and coordinate transform rounding errors. 

I chose not to make it a tunable parameter since it's a numerical guard tied strictly to map representation accuracy, rather than a design-level margin. The `min_drivable_area_margin` parameter already serves the purpose of a tunable safety boundary.

Fixed in [cbfe153](https://github.com/autowarefoundation/autoware_universe/pull/12068/commits/cbfe1536fbc20d79f3610401f40c89ccb6a69332).

File: planning/behavior_path_planner/autoware_behavior_path_side_shift_module/src/scene.cpp [hunk 3]
```diff
@@ -415,48 +514,223 @@ BehaviorModuleOutput SideShiftModule::adjustDrivableArea(const ShiftedPath & pat
...
+  }
+
+  if (!limits.found_valid_limit) {
+    RCLCPP_WARN_THROTTLE(
+      getLogger(), *clock_, 1000, "SideShift: No valid lane found for limit check. Returning 0.0");
+    return 0.0;
+  }
+
+  return clampOffsetToLimits(requested_offset, limits.safe_left_limit, limits.safe_right_limit);
+}
```
[the remark this answers is not in the document]
Comment (k-uta) [reply planning/behavior_path_planner/autoware_behavior_path_side_shift_module/src/scene.cpp:643]: Extracted the pure computational helpers (`determineLanePosition`, `updateLaneLimitsForOutsidePoint`, `updateLaneLimitsForInsidePoint`, `clampOffsetToLimits`) and their associated structs (`LaneLimitInfo`, `AdjacentLaneInfo`, `LanePositionResult`, `LaneCheckContext`) into a standalone `drivable_area_utils.{hpp,cpp}` module. This allows direct testing with synthetic lanelets, without requiring a ROS node or `planner_data_`.

Added **42 unit tests** in `test_drivable_area_utils.cpp` covering:

- **`clampOffsetToLimits`** (10 tests): zero request, within/exceeding limits for both left and right shifts, tolerance snapping to zero, exact-at-limit, zero-limit cases
- **`determineLanePosition`** (6 tests): point in current lane, left/right adjacent lane, outside all lanes, no adjacent lanes registered, boundary point
- **`updateLaneLimitsForOutsidePoint`** (4 tests): left/right side detection, return distance calculation, multi-point accumulation
- **`updateLaneLimitsForInsidePoint`** (8 tests): centered point, offset from center, narrow lane with insufficient space, left/right/both adjacent lane expansion, `point_in_adjacent` guard preventing further expansion, min-accumulation across multiple points
- **Integration scenarios** (5 tests): clamped left/right shifts, blocked narrow lanes, adjacent-lane expansion vs single-lane limits, mixed outside + inside point accumulation
- **Edge cases** (4 tests): boundary points, very large offsets, negative margins, zero-width vehicle
- **Default values and constants** (5 tests)

The remaining `calcMaxLateralOffset` orchestration (which depends on `planner_data_` and `route_handler`) and `getAdjacentLaneInfo` remain as `SideShiftModule` member functions since they require ROS-level dependencies. A follow-up refactor (5ac47ab) introduced `LaneCheckContext` to reduce function argument counts and a `LaneletGeometry` struct to reduce test setup duplication.

Fixed in d034cf63 and 5ac47ab.

File: planning/behavior_path_planner/autoware_behavior_path_side_shift_module/src/scene.cpp [hunk 4]
```diff
@@ -245,6 +278,26 @@ void SideShiftModule::replaceShiftLine()
 
   const auto new_sl = calcShiftLine();
 
+  // Check if the shift was aborted (indicated by zero-length shift with same start/end index)
+  const bool shift_aborted =
+    (new_sl.start_idx == new_sl.end_idx) && (std::abs(new_sl.end_shift_length) < 0.01);
```
[the remark this answers is not in the document]
Comment (k-uta) [reply planning/behavior_path_planner/autoware_behavior_path_side_shift_module/src/scene.cpp:287]: Thanks for catching this! I agree that the abort signal from `calcShiftLine` was incomplete.

Rather than relaxing the detection condition (which could false-positive on legitimately near-zero shifts), I've fixed the source: added `shift_line.end_idx = shift_line.start_idx;` in `calcShiftLine`'s abort branch so the signal is explicit and both conditions are satisfied.

Fixed in e89e1b1.

File: planning/behavior_path_planner/autoware_behavior_path_side_shift_module/src/scene.cpp [hunk 5]
```diff
@@ -415,48 +514,223 @@ BehaviorModuleOutput SideShiftModule::adjustDrivableArea(const ShiftedPath & pat
...
+  const double dist_to_right = lanelet::geometry::distance2d(right_bound, target_point);
+
+  const double extra_buffer = 0.05;
+
+  // Calculate maximum left shift (positive value)
+  double current_max_left = dist_to_left - vehicle_half_width - margin - extra_buffer;
+  if (allow_left && !point_in_adjacent) {
+    const double dist_to_far_left =
+      lanelet::geometry::distance2d(left_lane_val.leftBound2d(), target_point);
+    current_max_left = dist_to_far_left - vehicle_half_width - margin - extra_buffer;
```
[the remark this answers is not in the document]
Comment (k-uta) [reply planning/behavior_path_planner/autoware_behavior_path_side_shift_module/src/scene.cpp]: Thanks for the suggestion. While `getAdjacentLaneInfo` currently always sets `allow_left` and `has_left` together (so `allow_left == true` implies a valid lanelet), I agree that adding a defensive guard is good practice for maintainability. Fixed by adding `&& adj.has_left` / `&& adj.has_right` to the conditions.

Fixed in dcb2812.

File: planning/behavior_path_planner/autoware_behavior_path_side_shift_module/src/scene.cpp [hunk 6]
```diff
@@ -415,48 +514,223 @@ BehaviorModuleOutput SideShiftModule::adjustDrivableArea(const ShiftedPath & pat
...
-  {
-    extended_path.points.insert(
-      extended_path.points.end(), original_path.points.begin() + orig_ego_idx,
-      original_path.points.end());
+  // Calculate maximum right shift
+  double current_max_right = dist_to_right - vehicle_half_width - margin - extra_buffer;
+  if (allow_right && !point_in_adjacent) {
+    const double dist_to_far_right =
+      lanelet::geometry::distance2d(right_lane_val.rightBound2d(), target_point);
+    current_max_right = dist_to_far_right - vehicle_half_width - margin - extra_buffer;
```
[the remark this answers is not in the document]
Comment (k-uta) [reply planning/behavior_path_planner/autoware_behavior_path_side_shift_module/src/scene.cpp]: Same fix applied — added `&& adj.has_right` guard. See above.
